### PR TITLE
NFT Detail Screen

### PR DESCRIPTION
**Pull request by tbezman**
[**Loom PR Walkthrough**](https://www.loom.com/share/725fbf0271fa41ddbd6181373e47f1cd)

### Demo Video

https://user-images.githubusercontent.com/6754223/230135868-fead4ca6-56eb-4dc1-96fd-79810eeeeab8.mp4

### Screenshots
| 1 | 2 | 3 | 4 | 5 |
|--------|--------|--------|--------|--------|
| Simulator Screenshot - iPhone 14 - 2023-04-04 at 15 36 59 | Simulator Screenshot - iPhone 14 - 2023-04-04 at 15 36 50 | Simulator Screenshot - iPhone 14 - 2023-04-04 at 15 36 41 | Simulator Screenshot - iPhone 14 - 2023-04-04 at 15 36 24 | Simulator Screenshot - iPhone 14 - 2023-04-04 at 15 36 12

### Testing Locally
Head to `NftDetailScreen.tsx` and comment in one of these lines
image

### Callouts

1. We're not relying on dimensions from the server right now. The layout shift on the screen will get better once we start to use that as the source of truth for dimensions. It's not quite ready to rely on. Either way it's important that we have a client fallback to figure out the asset sizing with or without the server.
2. We're not supporting the following media types right now
  a. TextMedia
  b. GltfMedia
3. We haven't handled loading failures yet. I think we'll share some logic across web and mobile for that soon enough

### CODE REVIEW

File: apps/mobile/src/components/SvgWebView.tsx [hunk 1]
```diff
@@ -57,19 +57,14 @@ export function SvgWebView({ source, onLoadStart, onLoadEnd, style }: SvgWebView
   const [svgState, setSvgState] = useState<SvgContentState>({ kind: 'loading' });
 
   useEffect(() => {
-    const controller = new AbortController();
-    const signal = controller.signal;
-
     async function fetchSvg() {
       if (uri) {
-        onLoadStart?.();
```
Comment (tbezman) [inline apps/mobile/src/components/SvgWebView.tsx:65]: why did you remove this?

File: apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAsset.tsx [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      // This will fit the image to the screen appropriately.
+      const containerDimensions: Dimensions = { width: viewDimensions.width, height: MAX_HEIGHT };
+
+      return fitDimensionsToContainer({
+        container: containerDimensions,
+        source: imageState.dimensions,
+      });
+    }
+
+    return { width: 300, height: 300 };
```
Comment (tbezman) [inline apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAsset.tsx:99]: maybe just throw this away

File: apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAssetHtml.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+export function NftDetailAssetHtml({ onLoad, htmlUrl }: Props) {
+  return (
+    <View style={{ width: '100%', aspectRatio: 1 }}>
+      <WebView
+        onLoadEnd={onLoad}
+        originWhitelist={['*']}
+        className="h-full w-full bg-transparent"
+        scrollEnabled={false}
+        showsHorizontalScrollIndicator={false}
+        showsVerticalScrollIndicator={false}
```
Comment (tbezman) [inline apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAssetHtml.tsx:18]: do we want this to be interactive?

File: apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAssetVideo.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      style={outputDimensions}
+      onReadyForDisplay={(event) => {
+        onLoad(event.naturalSize);
+      }}
+      shouldPlay
+      isLooping
+      resizeMode={ResizeMode.CONTAIN}
+      source={{
+        uri: videoUrl,
+        headers: { Accepts: 'image/avif;image/png' },
```
Comment (tbezman) [inline apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAssetVideo.tsx:23]: remove these headers

File: apps/mobile/index.js [hunk 5]
```diff
@@ -8,6 +8,22 @@ import * as Sentry from 'sentry-expo';
...
+
+    if (joined.includes('The provided value')) {
+      return;
+    }
+
+    previousWarn(...args);
+  };
+}
+
+filterOutAnnoyingWarnings();
```
Comment (Robinnnnn) [inline apps/mobile/index.js:25]: lmao

File: apps/mobile/package.json [hunk 6]
```diff
@@ -41,6 +42,7 @@
     "react-native-tab-view": "^3.4.0",
     "react-native-webview": "^11.26.1",
     "sentry-expo": "~6.1.0",
+    "swr": "^2.1.1",
```
Comment (Robinnnnn) [inline apps/mobile/package.json:45]: swr is back 😍

File: apps/mobile/src/components/Pill.tsx [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { PropsWithChildren } from 'react';
+import { View, ViewProps } from 'react-native';
+
+type Props = PropsWithChildren<{ className?: string; style?: ViewProps['style'] }>;
+
+export function Pill({ children, style }: Props) {
```
Comment (Robinnnnn) [inline apps/mobile/src/components/Pill.tsx:6]: 😍

File: apps/mobile/src/components/SafeAreaViewWithPadding.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
+/**
+ * The goal of this component is to ensure that on phones with or without a notch,
+ * the content has some space from the top and bottom of the screen.
+ */
```
Comment (Robinnnnn) [inline apps/mobile/src/components/SafeAreaViewWithPadding.tsx:4]: this is so clutch

File: apps/mobile/src/screens/NftDetailScreen/NftAdditionalDetails.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+        ...NftAdditionalDetailsPOAPFragment
+        ...NftAdditionalDetailsTezosFragment
+        ...NftAdditionalDetailsEthFragment
+      }
+    `,
+    tokenRef
+  );
+
+  if (token.chain === 'POAP') {
```
Comment (Robinnnnn) [inline apps/mobile/src/screens/NftDetailScreen/NftAdditionalDetails.tsx:29]: great call

File: apps/mobile/src/screens/NftDetailScreen/NftAdditionalDetailsEth.tsx [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    return null;
+  }, [contract?.contractAddress?.address, tokenId]);
+
+  return (
+    <View className="flex flex-col space-y-4">
+      {token.owner?.username && (
+        <DetailSection>
+          <DetailLabelText>OWNED BY</DetailLabelText>
+
+          {/* TODO(Terence) When the user profile screen is ready, setup the onPress here */}
```
Comment (Robinnnnn) [inline apps/mobile/src/screens/NftDetailScreen/NftAdditionalDetailsEth.tsx:65]: you tracking this somewhere?

File: apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAsset.tsx [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+      // Width is the width of the parent view (the screen - some padding)
+      // Height is the max height for the image
+      //
+      // This will fit the image to the screen appropriately.
+      const containerDimensions: Dimensions = { width: viewDimensions.width, height: MAX_HEIGHT };
+
+      return fitDimensionsToContainer({
+        container: containerDimensions,
+        source: imageState.dimensions,
+      });
```
Comment (Robinnnnn) [inline apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAsset.tsx:96]: 🔥🔥🔥🔥🔥

File: apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAsset.tsx (hunk 2 above)
Comment (Robinnnnn) [reply apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAsset.tsx:102]: this may need to be dynamic if the image is displayed within a smaller container right?

File: apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAssetImage.tsx [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+type Props = {
+  imageUrl: string;
+  outputDimensions: Dimensions;
+  onLoad: (dimensions: Dimensions | null) => void;
+};
+
+export function NftDetailAssetImage({ imageUrl, outputDimensions, onLoad }: Props) {
+  if (imageUrl.includes('svg')) {
+    console.log(imageUrl);
```
Comment (Robinnnnn) [inline apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAssetImage.tsx:15]: keeping for now?

File: apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAssetVideo.tsx [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  onLoad: (dimensions: Dimensions) => void;
+};
+
+export function NftDetailAssetVideo({ outputDimensions, videoUrl, onLoad }: Props) {
+  return (
+    <Video
+      style={outputDimensions}
+      onReadyForDisplay={(event) => {
+        onLoad(event.naturalSize);
+      }}
```
Comment (Robinnnnn) [inline apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAssetVideo.tsx:17]: nice, thanks for figuring out the various onLoad callbacks

File: apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAsset.tsx [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    return { width: 300, height: 300 };
+  }, [imageState, viewDimensions, windowDimensions.height]);
+
+  const inner = useMemo(() => {
+    if (
+      token.media?.__typename === 'GIFMedia' ||
+      token.media?.__typename === 'ImageMedia' ||
+      token.media?.__typename === 'AudioMedia'
+    ) {
+      const imageUrl = token.media.previewURLs?.large;
```
Comment (Robinnnnn) [inline apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAsset.tsx:111]: let's talk about this

**Review by Robinnnnn — APPROVED** [review]

File: apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAsset.tsx (hunk 14 above)
Comment (tbezman) [reply apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAsset.tsx:111]: Resolved on the phone. We'll keep this for now until a user complains that their token isn't high quality enough.

File: apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAssetImage.tsx (hunk 12 above)
Comment (tbezman) [reply apps/mobile/src/screens/NftDetailScreen/NftDetailAsset/NftDetailAssetImage.tsx:15]: Removed.

File: apps/mobile/src/screens/NftDetailScreen/NftAdditionalDetailsEth.tsx (hunk 10 above)
Comment (tbezman) [reply apps/mobile/src/screens/NftDetailScreen/NftAdditionalDetailsEth.tsx:65]: https://linear.app/galleryso/issue/GAL-2782/wire-up-links-from-other-screens
